packages/eslint-plugin/rules/no-mixed-plugin-imports.js

```diff
@@ -0,0 +1,160 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// @ts-check
+
+const visitImports = require('../lib/visitImports');
+const getPackages = require('../lib/getPackages');
+const minimatch = require('minimatch');
+
+/** @typedef {import('../lib/getPackages.js').ExtendedPackage} ExtendedPackage */
+
+/**
+ * @param {string} pattern
+ * @param {string} filePath
+ * @returns {boolean}
+ */
+const matchesPattern = (pattern, filePath) => {
+  return new minimatch.Minimatch(pattern).match(filePath);
+};
+
+const roleRules = [
+  {
+    sourceRole: ['frontend-plugin', 'web-library'],
+    targetRole: [
+      'backend-plugin',
+      'node-library',
+      'backend-plugin-module',
+      'frontend-plugin',
+    ],
+  },
+  {
+    sourceRole: ['backend-plugin', 'node-library', 'backend-plugin-module'],
+    targetRole: ['frontend-plugin', 'web-library', 'backend-plugin'],
+  },
+  {
+    sourceRole: ['common-library'],
+    targetRole: [
+      'frontend-plugin',
+      'web-library',
+      'backend-plugin',
+      'node-library',
+      'backend-plugin-module',
+    ],
+  },
+];
+
+/** @type {import('eslint').Rule.RuleModule} */
+module.exports = {
+  meta: {
+    type: 'problem',
+    messages: {
+      forbidden:
+        '{{sourcePackage}} ({{sourceRole}}) uses forbidden import from {{targetPackage}} ({{targetRole}}).',
+    },
+    docs: {
+      description: 'Disallow mixed plugin imports.',
+      url: 'https://github.com/backstage/backstage/blob/master/packages/eslint-plugin/docs/rules/no-mixed-plugin-imports.md',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          excludedTargetPackages: {
+            type: 'array',
+            items: { type: 'string' },
+            uniqueItems: true,
+          },
+          excludedFiles: {
+            type: 'array',
+            items: { type: 'string' },
+            uniqueItems: true,
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+  create(context) {
+    const packages = getPackages(context.cwd);
+    if (!packages) {
+      return {};
+    }
+
+    const filePath = context.physicalFilename
+      ? context.physicalFilename
+      : context.filename;
+
+    /** @type {ExtendedPackage | undefined} */
+    const pkg = packages.byPath(filePath);
+    if (!pkg) {
+      return {};
+    }
+
+    const options = context.options[0] || {};
+    /** @type {string[]} */
+    const ignoreTargetPackages = options.excludedTargetPackages || [];
+    /** @type {string[]} */
+    const ignorePatterns = options.excludedFiles || [
+      '**/*.{test,spec}.[jt]s?(x)',
```

Comment:
Test files are ignored by default as for example `startTestBackend` requires to import the actual plugin, f.eg:

```
Lint failed in plugins/auth-backend-module-oidc-provider

✘  https://google.com/#q=%40backstage%2Fno-mixed-plugin-imports

     @backstage/plugin-auth-backend-module-oidc-provider (backend-plugin-module) uses forbidden import from @backstage/plugin-auth-backend (backend-plugin)


     src/module.test.ts:147:9
     145 |       features: [
     146 |         authModuleOidcProvider,
   > 147 |         import('@backstage/plugin-auth-backend'),
         |         ^
     148 |         mockServices.rootConfig.factory({
     149 |           data: {
     150 |             app: { baseUrl: 'http://localhost' },
```